nestkernel/eprop_archiving_node.h

```diff
@@ -117,21 +199,99 @@ class EpropArchivingNodeRecurrent : public EpropArchivingNode< HistEntryEpropRec
+        { "fast_sigmoid_derivative", &EpropArchivingNodeRecurrent::compute_fast_sigmoid_derivative_surrogate_gradient },
+        { "arctan", &EpropArchivingNodeRecurrent::compute_arctan_surrogate_gradient } };
+
+    auto found_entry_it = surrogate_gradient_funcs.find( surrogate_gradient_function );
+    if ( found_entry_it == surrogate_gradient_funcs.end() )
+    {
+      std::string error_message = "Surrogate gradient / pseudo-derivate function surrogate_gradient_function from [";
+      for ( const auto& surrogate_gradient_func : surrogate_gradient_funcs )
+      {
+        error_message += " \"" + surrogate_gradient_func.first + "\",";
+      }
+      error_message.pop_back();
+      error_message += " ] required.";
+
+      throw BadProperty( error_message );
+    }
+    else
+    {
+      return found_entry_it->second;
+    }
+  }
+
+  //! Compute the surrogate gradient with a piecewise linear function around the spike time (used, e.g., in Bellec et
+  //! al., 2020).
+  double compute_piecewise_linear_surrogate_gradient( const double r,
+    const double v_m,
+    const double v_th_adapt,
+    const double V_th,
+    const double beta,
+    const double gamma );
+
+  //! Compute the surrogate gradient with an exponentially decaying function around the spike time (used, e.g., in
+  //! Shrestha and Orchard, 2018).
+  double compute_exponential_surrogate_gradient( const double r,
+    const double v_m,
+    const double v_th_adapt,
+    const double V_th,
+    const double beta,
+    const double gamma );
+
+  //! Compute the surrogate gradient with a function corresponding to the derivative of a fast sigmoid around the spike
+  //! (used, e.g., in Zenke and Ganguli, 2018).
+  double compute_fast_sigmoid_derivative_surrogate_gradient( const double r,
+    const double v_m,
+    const double v_th_adapt,
+    const double V_th,
+    const double beta,
+    const double gamma );
+
+  //! Compute the surrogate gradient with an arctan function around the spike time (used, e.g., in Fang et al., 2021).
+  double compute_arctan_surrogate_gradient( const double r,
+    const double v_m,
+    const double v_th_adapt,
+    const double V_th,
+    const double beta,
+    const double gamma );
+
+  //! Create an entry for the given time step at the end of the eprop history.
+  void emplace_new_eprop_history_entry( const long time_step );
+
+  //! Write the given surrogate gradient value to the history at the given time step.
   void write_surrogate_gradient_to_history( const long time_step, const double surrogate_gradient );
 
   //! Update the learning signal in the eprop history entry of the given time step by writing the value of the incoming
   //! learning signal to the history or adding it to the existing value in case of multiple readout neurons.
-  void write_learning_signal_to_history( const long time_step, const double learning_signal );
+  void write_learning_signal_to_history( const long time_step,
```

Comment:
Could you document parameters in the comment, especially `has_norm_step`?

Reply:
Fixed in https://github.com/nest/nest-simulator/pull/3207/commits/102dc9099b459097152dafa06ccedb18a1ad113b